### PR TITLE
D-ID Tool

### PR DESCRIPTION
**Pull request by eob**
Putting this up here to scaffold a conversation about an easy V1 packaging

### CODE REVIEW

File: src/steamship/agents/tools/video_veneration/did_video_generator_tool.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+from steamship.agents.tools import ImageGeneratorTool
+from steamship.utils.repl import ToolREPL
+
+
+class DIDVideoGeneratorOptions(BaseModel):
+    source_url: str
+    """The URL of the source image to be animated."""
+
+    stitch: str
```
Comment (eob) [inline src/steamship/agents/tools/video_veneration/did_video_generator_tool.py:14]: @EniasCailliau I wasn't sure what this param in the plugin config referred to

File: src/steamship/agents/tools/video_veneration/did_video_generator_tool.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    provider: Optional[str]
```
Comment (eob) [inline src/steamship/agents/tools/video_veneration/did_video_generator_tool.py:16]: Ditto here re being unsure (`provider`)

File: src/steamship/agents/tools/video_veneration/did_video_generator_tool.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    """The URL of the D-ID driver video. If not provided a driver video will be selected automatically."""
+
+    driver_url: Optional[str]
+    """The URL of the D-ID driver video. If not provided a driver video will be selected automatically."""
+
+    expressions: Optional[List[dict]]
```
Comment (eob) [inline src/steamship/agents/tools/video_veneration/did_video_generator_tool.py:22]: @EniasCailliau is this the right format for the `expressions` field? 

The `talk` generation API docs made it seem as if this was an array of frame-pegged markers, but that seemed to require knowledge of the frames

File: src/steamship/agents/tools/video_generation/__init__.py [hunk 4]
```diff
@@ -0,0 +1,5 @@
+from .did_video_generator_tool import DIDVideoGeneratorTool
+
+__all__ = [
+    "DIDVideoGeneratorTool",
```
Comment (douglas-reid) [inline src/steamship/agents/tools/video_generation/__init__.py:4]: random nonsense aside: i find `DID` disconcerting. i would like to speak to their branding folks!

File: src/steamship/agents/tools/video_generation/did_video_generator_tool.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
+"""Tool for generating images."""
+from typing import Any, List, Optional, Union
+
+from steamship import Block, Task
+from steamship.agents.schema import AgentContext
+from steamship.agents.tools import VideoGeneratorTool
+from steamship.utils.repl import ToolREPL
+
+DEFAULT_SOURCE_URL = "https://www.steamship.com/images/agents/man-in-suit-midjourney.png"
```
Comment (douglas-reid) [inline src/steamship/agents/tools/video_generation/did_video_generator_tool.py:9]: i don't know how to calibrate, etc., in the right way here. `man in suit` seems perfectly reasonable. but i wonder if there will be any emotional baggage with this default choice.

File: src/steamship/agents/tools/video_generation/did_video_generator_tool.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    )
+    generator_plugin_handle: str = "did-video-generator"
+    generator_plugin_config: dict = {}
+
+    source_url: Optional[str] = DEFAULT_SOURCE_URL
+    """The URL of the source image to be animated."""
+
+    stitch: bool = True
+
+    voice_provider: Optional[str] = "microsoft"
```
Comment (douglas-reid) [inline src/steamship/agents/tools/video_generation/did_video_generator_tool.py:30]: @eob should we make this a `(Str,Enum)` then?

File: src/steamship/agents/tools/video_generation/did_video_generator_tool.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    """The voice provider. Must be either `microsoft` or `amazon`."""
+
+    voice_id: Optional[str] = "en-US-GuyNeural"
```
Comment (douglas-reid) [inline src/steamship/agents/tools/video_generation/did_video_generator_tool.py:33]: will this eventually be independent, or is this entirely dependent on `voice_provider`?

File: src/steamship/agents/tools/video_generation/did_video_generator_tool.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    """The voice ID. E.g. `en-US-AshleyNeural` for Microsoft or `Amy` for Amazon."""
+
+    driver_url: Optional[str] = None
+    """The URL of the D-ID driver video. If not provided a driver video will be selected automatically."""
+
+    expressions: Optional[List[dict]] = None
+    """A list of expressions to apply.
```
Comment (douglas-reid) [inline src/steamship/agents/tools/video_generation/did_video_generator_tool.py:40]:
```suggestion
    """A list of expressions (and their intensity) to apply.
```

File: src/steamship/agents/tools/video_generation/did_video_generator_tool.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+    Valid expressions are: neutral | happy | surprise | serious
+    Intensity is a float from 0 to 1.0
+
+    Use the following form
+
+    [
+        {
```
Comment (douglas-reid) [inline src/steamship/agents/tools/video_generation/did_video_generator_tool.py:48]: is this a set schema?  should we make a type to represent it (and allow validation, etc.)?

File: src/steamship/agents/tools/video_generation/did_video_generator_tool.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+            "start_frame": 0,
+            "expression": "surprise",
+            "intensity": 1.0
+        },
+    ]
```
Comment (douglas-reid) [inline src/steamship/agents/tools/video_generation/did_video_generator_tool.py:53]: what happens if i have multiple expressions with the same `start_frame` ?  does order in the list matter, or does DID handle that based on `start_frame`?

File: src/steamship/agents/tools/video_generation/did_video_generator_tool.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        "The input is the text that you want to say. "
+        "The output is the video of you saying it."
+    )
+    generator_plugin_handle: str = "did-video-generator"
+    generator_plugin_config: dict = {}
+
+    source_url: Optional[str] = DEFAULT_SOURCE_URL
+    """The URL of the source image to be animated."""
+
+    stitch: bool = True
```
Comment (douglas-reid) [inline src/steamship/agents/tools/video_generation/did_video_generator_tool.py:28]: what is the implication of setting this to `False` ?

File: src/steamship/agents/tools/video_generation/did_video_generator_tool.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+        if self.merge_blocks:
+            block = tool_input[0]
+            for extra_block in tool_input[1:]:
+                block.text = f"{block.text}\n\n{extra_block.text}"
+            tool_input = [block]
+
+        for block in tool_input:
+            if isinstance(block, str):
+                print(block)
```
Comment (douglas-reid) [inline src/steamship/agents/tools/video_generation/did_video_generator_tool.py:78]: is this entire bit about debugging?  should it be kept in?

File: src/steamship/agents/tools/video_generation/did_video_generator_tool.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+            if not block.is_text():
+                continue
+
+            prompt = block.text
+            task = generator.generate(
+                text=prompt,
+                make_output_public=True,
```
Comment (douglas-reid) [inline src/steamship/agents/tools/video_generation/did_video_generator_tool.py:85]: this _feels_ like something that should be configurable. thoughts?

**Review by douglas-reid — COMMENTED** [review]
took a quick pass. hopefully asked some useful questions. in general, excited about this.

File: src/steamship/agents/tools/video_veneration/did_video_generator_tool.py [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+
+from pydantic import BaseModel
+
+from steamship.agents.tools import ImageGeneratorTool
+from steamship.utils.repl import ToolREPL
+
+
+class DIDVideoGeneratorOptions(BaseModel):
+    source_url: str
+    """The URL of the source image to be animated."""
```
[the remark this answers is not in the document]
Comment (EniasCailliau) [reply src/steamship/agents/tools/video_veneration/did_video_generator_tool.py]: +1

File: src/steamship/agents/tools/video_veneration/did_video_generator_tool.py (hunk 1 above)
Comment (EniasCailliau) [reply src/steamship/agents/tools/video_veneration/did_video_generator_tool.py]: This should be a bool. Will push a fix.

File: src/steamship/agents/tools/video_veneration/did_video_generator_tool.py (hunk 2 above)
Comment (EniasCailliau) [reply src/steamship/agents/tools/video_veneration/did_video_generator_tool.py]: Going to add the BaseModel's I've designed for them.

File: src/steamship/agents/tools/video_veneration/did_video_generator_tool.py (hunk 3 above)
Comment (EniasCailliau) [reply src/steamship/agents/tools/video_veneration/did_video_generator_tool.py]: Yea. It's not ideal. To make it worse they raise exceptions when you referenced non-existing markers.

**Review by EniasCailliau — APPROVED** [review]

**Review by EniasCailliau — APPROVED** [review]

**Review by EniasCailliau — APPROVED** [review]

**Review by EniasCailliau — APPROVED** [review]

**Review by EniasCailliau — APPROVED** [review]

File: src/steamship/agents/tools/video_generation/did_video_generator_tool.py (hunk 13 above)
Comment (eob) [reply src/steamship/agents/tools/video_generation/did_video_generator_tool.py:101]: Agree -- I tried for formulate an interface for it, but it didn't come out confident-feeling enough to add to this PR. 

The two I tried were `get_generate_args` and `do_gnerate` on the base class.

File: src/steamship/agents/tools/video_generation/did_video_generator_tool.py (hunk 12 above)
Comment (eob) [reply src/steamship/agents/tools/video_generation/did_video_generator_tool.py:94]: Nice catch!

File: src/steamship/agents/tools/video_generation/did_video_generator_tool.py (hunk 11 above)
Comment (eob) [reply src/steamship/agents/tools/video_generation/did_video_generator_tool.py:42]: No clue. It's on the D-ID plugin so I added it here to match. For this and the comments about the other config fields, I'm going to wait until V2 to address; agree with them all but it sounds like @EniasCailliau has a config object that he's working on that we might be able to just use directly for a 1:1 documented, pydantic-friendly obj

**Pull request merged**